packages/@aws-cdk/aws-codepipeline-actions/lib/s3/deploy-action.ts

```diff
@@ -106,11 +107,18 @@ export class S3DeployAction extends Action {
     this.props = props;
   }
 
-  protected bound(_scope: Construct, _stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
+  protected bound(scope: Construct, _stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
   codepipeline.ActionConfig {
     // pipeline needs permissions to write to the S3 bucket
     this.props.bucket.grantWrite(options.role);
 
+    if (this.props.accessControl !== undefined &&
+        FeatureFlags.of(scope).isEnabled(cxapi.S3_GRANT_WRITE_WITHOUT_ACL)) {
```

Comment:
This feels worse because it's tying implementation details of two different classes together.

To my mind, we should be replacing

```ts
this.props.bucket.grantWrite(options.role);
```

With either:

```ts
this.props.bucket.grant(options.role,
   's3:PutObject',
   's3:PubObject*Acl',
   's3:DeleteObject',
   's3:...WhateverElseHere
);
```

OR

```ts
this.props.bucket.grantWriteWithAcl(options.role);
```

CDK methods should be *intent-based*. You tell me what you need, I'll make it happen. 

Not: "you tell me be a bit of what you need, and you'll predict what I'll have done based on some flags, and then you tell me to do some more"

Reply:
I don't like either of those options 😕.

`this.props.bucket.grant(options.role, 's3:PutObject', 's3:PubObject*Acl', ...)` I think is out of the question - it's way too low-level, and completely not intent-based.

I think `grantWriteWithAcl()` would actually be a mistake. One of the problems that were raised with `grantWrite()` giving the permissions it does now is that `s3:PutObjectAcl` is not a write permissions - it's a permission-changing permission, and that it's surprising that it's present on a method that deals with writing. So, having `grantWriteWithAcl()` encourages a practice that our security team considers bad (mixing permissions in different categories).

Let's just always call `bucket.grantPutAcl()` if `accessControl` was set, regardless of the presence of the feature flag. The code will be simple and, like you said, intent-based, and we'll just have superfluous permissions in the case the feature flag is not set.

I've changed the code to match, let me know if this works!